### PR TITLE
[monorepo] feat: checkout submodule

### PR DESCRIPTION
**Pull request by Wayonb**
Linters and Jenkins folder is shared across monorepos.  The Jenkins job needs to check out submodule to get access to the Dockerfile to run the pipeline.

### CODE REVIEW

File: jenkins/shared-library/vars/defaultCiPipeline.groovy [hunk 1]
```diff
@@ -298,3 +299,11 @@ String resolvePath(String rootPath, String path) {
 Boolean shouldPublishImage(String shouldPublish) {
 	return shouldPublish == null ? false : shouldPublish.toBoolean()
 }
+
+void runInitializeScriptIfPresent() {
+	String initFile = 'init.sh'
+	if (fileExists(initFile)) {
+		println 'Running initialized script'
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/defaultCiPipeline.groovy:306]: initialize

File: jenkins/shared-library/vars/createMultibranchJob.groovy [hunk 2]
```diff
@@ -43,6 +44,39 @@ void call(Map jobConfiguration) {
...
 									}
+
+									cleanBeforeCheckoutTrait {
+										extension {
+											// Deletes untracked submodules and any other subdirectories which contain .git directories.
+											deleteUntrackedNestedRepositories(true)
+										}
+									}
+
+									submoduleOptionTrait {
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/createMultibranchJob.groovy:55]: why is this needed ?
is this cloning the submodules ?

Comment (Wayonb) [reply jenkins/shared-library/vars/createMultibranchJob.groovy:55]: Yea, Jenkins needs access to the Dockerfiles in ``jenkins/docker`` to build the image before the stages in pipeline. I needed to clone the submodules.

Comment (Jaguar0625) [reply jenkins/shared-library/vars/createMultibranchJob.groovy:55]: is there a way to only clone a subset of the repo using sparsecheckout, like init.sh is doing?

Comment (Wayonb) [reply jenkins/shared-library/vars/createMultibranchJob.groovy:55]: I was looking last night but havent found a way yet.  Will continue looking today.  Hopefully with fresh eyes might have better luck :)

Comment (Wayonb) [reply jenkins/shared-library/vars/createMultibranchJob.groovy:55]: I havent found a good way to checkout the submodule using sparsecheckout.  The git plug doesnt support it for submodules(https://issues.jenkins.io/browse/JENKINS-42628)
Workaround would be
1. Use git hooks(post-checkout) which runs the init.sh
2. In the pipeline, create a new stage that will run init.sh script before the Dockerfile.  This will move the Dockerfile into the stages
3. Could probably enable ``Shallow clone`` to save some time and space.
@Jaguar0625 @gimre-xymcity any thoughts?

Comment (Jaguar0625) [reply jenkins/shared-library/vars/createMultibranchJob.groovy:55]: i think this is ok for now, but you can try shallow clone and see if it improves things. if so, you can make another PR with those changes. 👍

**Review by Jaguar0625 — APPROVED** [review]
